### ISSUE TITLE
skopeo login, with no XDG_RUNTIME_DIR, fails in a bad way

### ISSUE THREAD

**Issue by edsantiago** [containers/skopeo#1130]
Similar to #923 

```console
$ skopeo login quay.io
Username: myusername
Password:
FATA[0021] mkdir /run/containers/1000: permission denied
$ id -u
1000
```
`podman login`, however, works just fine:
```console
$ podman login quay.io
Username: myusername
Password:
Login Succeeded!
```
If I `export XDG_RUNTIME_DIR=/run/user/1000`, skopeo can use the credentials written by podman. I kind of expected skopeo and podman to assume the same defaults in an XDG-less environment.

TL;DR podman has a reasonable default for XDG_RUNTIME_DIR; skopeo should too.

**Comment by mtrmac** [comment]
Minimal summary:
- Why is XDG_RUNTIME_DIR empty?
- Either we fall back to the home directory (which @rhatdan intentionally wanted us to move away from), or there is nothing that works.

---


Philosophically, just for basic cross-application consistency reasons, it’s wrong for an end-user application that is not in charge of setting up the session, to override XDG_* environment variable values (or alternatively, it’s wrong for a library inside such an application to read XDG_* environment variable values, I guess; but more the former).

It just doesn’t work to have to keep N statically-linked executables that embed undocumented defaults in sync (when these defaults even change over time! see https://github.com/containers/podman/pull/8241 ).

Podman kind of has to set the variable because it manipulates user namespaces, but that doesn’t make it a viable feature.

So Skopeo just _can’t_ be consistent with Podman’s fallback shen^W…strategies as they change over time.

---

> If I `export XDG_RUNTIME_DIR=/run/user/1000`

Note that both `/run` and `/run/user` are (or better be) 0755 root:root ; so we can’t just default to those paths without some root-owned process creating those paths for us; and then my question would be why doesn’t that root-owned process set up the session to contain a valid `XDG_RUNTIME_DIR` (and there can be, in turn, some some complex answers that, incl. https://bugzilla.redhat.com/show_bug.cgi?id=753882 leading to sad things like https://cgit.freedesktop.org/polkit/commit/?id=8635ffc16aeff6a07d675f861fe0dea03ea81d7e )

There isn’t any default value that can be used; https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html  basically shrugs and says “find some other place”.

---

Apart from XDG_RUNTIME_DIR, focusing on credential storage, this is _kind of_ intentional: There was an intentional decision not to store credentials in the home directory — and, well, without the home directory we have nothing to fall back on, per the above.

Lots of history:
- https://github.com/containers/image/pull/424#pullrequestreview-101703066
- https://github.com/containers/image/pull/729
- https://github.com/containers/podman/issues/427#issuecomment-371221658
- https://github.com/containers/storage/pull/568

tl;dr the long-term hope was that we are going to migrate users to credential helpers, rather than default to home directories. That hasn’t happened yet (and for the contemplated user running in a stripped-down environment where not even XDG_* environment variables are set, it seems dubious to assume that credential helpers will be available).

**Comment by edsantiago** [comment]
`XDG_*` are empty because this is a session created via `su - $user`. This is not uncommon. The `/run/user/1000` directory exists because I have run (as root) `loginctl enable-linger $user`; this is a prerequisite for being able to use podman. 

I accept that there are unclear standards about XDG; and I'm an old neckbeard who doesn't understand or use or want any of these newfangled complications anyway (but that battle is lost). All I'm saying is: podman has one default, skopeo has another. The skopeo one is clearly wrong, because it throws an error. It would be nice if the skopeo default was the same as the podman one. (And yes, universal consensus that the default should not be $HOME).

**Comment by mtrmac** [comment]
Looking again, I’ve been a bit too harsh on Podman: its first fallback is `/run/user/$UID`, which makes much more sense than the other two.

----

> All I'm saying is: podman has one default, skopeo has another. The skopeo one is clearly wrong, because it throws an error. 

The Podman one is less clearly wrong… see elsewhere.

> It would be nice if the skopeo default was the same as the podman one.

I mean, yes, and we’ve been through a similar exercise with `--authfile`: the way to do that is to move the default inside the c/image library (and that default right now is the Skopeo one that fails, sure, and that probably needs fixing), and for Podman not to set a Podman-private override, let alone an override _that changes over time_.

---

Pragmatically, the `systemd` vs. `su` vs. `XDG_RUNTIME_DIR` situation will probably never improve, but it will also not _change_, so it might well make sense to (break existing login sessions once and) change the default path in c/image (_not_ Skopeo) to `/run/user/$EUID` from `/run/containers/$EUID` (except for `root`??), i.e. basically to assume systemd is managing the sessions, and we “only” lost `XDG_RUNTIME_DIR`.

That would allow `skopeo login`+`skopeo copy` to work in such environments by default (i.e. a host with systemd running). And that’s what Podman happens to do in such environments AFAICS.

In the other environments where Podman falls back from `/run/user` to other paths… (notably rootless use inside a container), the previous comment applies: I don’t know of any other good place with the required properties, and of a maintainable way to keep up with Podman changes.

---

I’ll move this to c/image, I don’t think Skopeo should be in that business (OTOH c/image shouldn’t be creating directories in unexpected places, so that move might be premature — we can move back if necessary).
